politeiad/backendv2/tstorebe/plugins/pi/hooks.go

```diff
@@ -0,0 +1,292 @@
+				}
+			}
+
+			// Verify text file size
+			if len(payload) > int(p.textFileSizeMax) {
+				e := fmt.Sprintf("file %v size %v exceeds max size %v",
+					v.Name, len(payload), p.textFileSizeMax)
+				return backend.PluginError{
+					PluginID:     pi.PluginID,
+					ErrorCode:    uint32(pi.ErrorCodeTextFileSizeInvalid),
+					ErrorContext: e,
+				}
+			}
+
+		case mimeTypePNG:
+			imagesCount++
+
+			// Verify image file size
+			if len(payload) > int(p.imageFileSizeMax) {
+				e := fmt.Sprintf("image %v size %v exceeds max size %v",
+					v.Name, len(payload), p.imageFileSizeMax)
+				return backend.PluginError{
+					PluginID:     pi.PluginID,
+					ErrorCode:    uint32(pi.ErrorCodeImageFileSizeInvalid),
+					ErrorContext: e,
+				}
+			}
+
+		default:
+			return fmt.Errorf("invalid mime: %v", v.MIME)
+		}
+	}
+
+	// Verify that an index file is present
+	var found bool
+	for _, v := range files {
+		if v.Name == pi.FileNameIndexFile {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return backend.PluginError{
+			PluginID:     pi.PluginID,
+			ErrorCode:    uint32(pi.ErrorCodeTextFileMissing),
+			ErrorContext: pi.FileNameIndexFile,
+		}
+	}
+
+	// Verify image file count is acceptable
+	if imagesCount > p.imageFileCountMax {
+		e := fmt.Sprintf("got %v image files, max is %v",
+			imagesCount, p.imageFileCountMax)
+		return backend.PluginError{
+			PluginID:     pi.PluginID,
+			ErrorCode:    uint32(pi.ErrorCodeImageFileCountInvalid),
+			ErrorContext: e,
+		}
+	}
+
+	// Verify a proposal metadata has been included
+	pm, err := proposalMetadataDecode(files)
+	if err != nil {
+		return err
+	}
+	if pm == nil {
+		return backend.PluginError{
+			PluginID:     pi.PluginID,
+			ErrorCode:    uint32(pi.ErrorCodeTextFileMissing),
+			ErrorContext: pi.FileNameProposalMetadata,
+		}
+	}
+
+	// Verify proposal name
+	if !p.proposalNameIsValid(pm.Name) {
+		return backend.PluginError{
+			PluginID:     pi.PluginID,
+			ErrorCode:    uint32(pi.ErrorCodeProposalNameInvalid),
+			ErrorContext: p.proposalNameRegexp.String(),
+		}
+	}
+
+	return nil
+}
+
+func (p *piPlugin) hookNewRecordPre(payload string) error {
+	var nr plugins.HookNewRecordPre
+	err := json.Unmarshal([]byte(payload), &nr)
+	if err != nil {
+		return err
+	}
+
+	return p.proposalFilesVerify(nr.Files)
+}
+
+func (p *piPlugin) hookEditRecordPre(payload string) error {
+	var er plugins.HookEditRecord
+	err := json.Unmarshal([]byte(payload), &er)
+	if err != nil {
+		return err
+	}
+
+	// Verify proposal files
+	err = p.proposalFilesVerify(er.Files)
+	if err != nil {
+		return err
+	}
+
+	// Verify vote status allows proposal edits
+	if er.RecordMetadata.State == backend.StateVetted {
+		t, err := tokenDecode(er.RecordMetadata.Token)
+		if err != nil {
+			return err
+		}
+		s, err := p.voteSummary(t)
+		if err != nil {
+			return err
+		}
+		if s.Status != ticketvote.VoteStatusUnauthorized {
+			e := fmt.Sprintf("vote status '%v' does not allow for proposal edits",
```

Comment:
Do this directly in the struct.

Reply:
All of these have been fixed throughout politeiad.